File: new-resource-detector/src/test/java/nl/haploid/resource/detector/service/ResourceDetectorsServiceIT.java
package nl.haploid.resource.detector.service;

import mockit.Deencapsulation;
import mockit.StrictExpectations;
import nl.haploid.event.RowChangeEvent;
import nl.haploid.resource.detector.AbstractIT;
import nl.haploid.resource.detector.TestData;
import org.junit.Before;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.annotation.DirtiesContext;

import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.assertEquals;

@DirtiesContext
public class ResourceDetectorsServiceIT extends AbstractIT {

	@Autowired
	private ResourceDetectorsService service;

	@Autowired
	private MockResourceDetector mockDetector;

	@Before
	public void setUp() throws Exception {
		Deencapsulation.setField(service, "detectors", Arrays.asList(mockDetector));
	}

	@Test
	public void testGetDetectorsForTable() {
		final List<MockResourceDetector> expectedDetectors = Arrays.asList(mockDetector);
		final String tableName = "kinsey";
		new StrictExpectations(mockDetector) {{
			mockDetector.getTableNames();
			times = 1;
			result = Arrays.asList("crane", tableName);
		}};
		final List<ResourceDetector> actualDetectors = service.getDetectorsForTable(tableName);
		assertEquals(expectedDetectors, actualDetectors);
	}

	@Test
	public void testDetectResources() {
		final String tableName = "harris";
		final RowChangeEvent event1 = TestData.rowChangeEvent(tableName);
		final RowChangeEvent event2 = TestData.rowChangeEvent("calvet");
		final List<RowChangeEvent> events = Arrays.asList(event1, event2);
		final List<ResourceDescriptor> expectedDescriptors = Arrays.asList(TestData.resourceDescriptor(96l));
		new StrictExpectations(mockDetector) {{
			mockDetector.getTableNames();
			times = 1;
			result = Arrays.asList("holloway", tableName);
			mockDetector.detect(events);
			times = 1;
			result = expectedDescriptors;
		}};
		final List<ResourceDescriptor> actualDescriptors = service.detectResources(tableName, events);
		assertEquals(expectedDescriptors, actualDescriptors);
	}
}

File: new-resource-detector/src/test/scala/nl/haploid/octowight/channel/scout/service/ResourceDetectorsServiceIT.scala
package nl.haploid.octowight.channel.scout.service

import java.util

import nl.haploid.octowight.channel.scout.detector.ResourceDetector
import nl.haploid.octowight.channel.scout.{AbstractIT, TestData}
import nl.haploid.octowight.{AtomGroup, EasyMockInjection, Mocked}
import org.scalatest.mock.EasyMockSugar
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.util.ReflectionTestUtils

@DirtiesContext
class ResourceDetectorsServiceIT extends AbstractIT with EasyMockSugar with EasyMockInjection {
  @Autowired private[this] val resourceDetectorsService: ResourceDetectorsService = null
  @Mocked private[this] val mockDetector: ResourceDetector = null

  behavior of "Resource detectors service"

  override def beforeEach() = {
    super.beforeEach()
    injectMocks()
    ReflectionTestUtils.setField(resourceDetectorsService, "_detectors", util.Collections.singletonList(mockDetector))
  }

  it should "get detectors for atom category" in {
    val expectedDetectors = Set(mockDetector)
    val atomCategory = "kinsey"
    expecting {
      mockDetector.atomCategories andReturn Set("crane", atomCategory) once()
    }
    whenExecuting(mockDetector) {
      val actualDetectors = resourceDetectorsService.detectorsForAtomCategory(atomCategory)
      actualDetectors should be(expectedDetectors)
    }
  }

  it should "detect resources" in {
    val atomCategory = "harris"
    val event1 = TestData.atomChangeEvent(atomCategory)
    val event2 = TestData.atomChangeEvent("calvet")
    val atomGroup = new AtomGroup(origin = event1.origin, category = event1.category)
    val atomIds = Set(event1.id, event2.id)
    val expectedResourceRoots = Set(TestData.resourceRoot(TestData.nextLong))
    expecting {
      mockDetector.atomCategories andReturn Set("holloway", atomCategory) once()
      mockDetector.detect(atomGroup, atomIds) andReturn expectedResourceRoots once()
    }
    whenExecuting(mockDetector) {
      val actualResourceRoots = resourceDetectorsService.detectResources(atomGroup, atomIds)
      actualResourceRoots should be(expectedResourceRoots)
    }
  }
}
